File: runtimes/web/src/core/ir/IRModels.ts
/**
 * Core IR (Intermediate Representation) types for SDUI — IR v2 wire shape.
 *
 * These types mirror the Kotlin IR models (converter IRWireV2.kt) and the
 * normative wire contract in schema/spec/01-envelope.md + 03-children.md:
 * a FLAT component list where composition is expressed by the child-side
 * `slot` reference, never by nesting. Legacy v1 documents (nested
 * `children`, `_text`/`_tag`/`_pseudo` underscore fields) are accepted for
 * one deprecation window via the decode gate in IRDecode.ts, which
 * translates them into this v2 shape so downstream code sees ONE model.
 */

/**
 * Root document envelope (schema/spec/01-envelope.md).
 *
 * `irVersion` states what the writer produced; `minReaderVersion` the
 * oldest reader that can safely consume it — both `2` on today's wire.
 * They are optional in the TYPE only because a raw fetch of a legacy v1
 * document has neither; after `decodeIRDocument` both are always 2.
 */
export interface IRDocument {
  /** Writer version. 2 on the v2 wire; absent on a raw legacy v1 doc. */
  irVersion?: number;
  /** Oldest reader that may consume this doc (reader refuses if > 2). */
  minReaderVersion?: number;
  /** FLAT component list — every component at every composition depth. */
  components: IRComponent[];
  /**
   * Document-level named keyframe sets (schema/spec/07-animations.md §1.2)
   * — the wire twin of CSS `@keyframes` at-rules, which are document-scoped
   * in CSS too. ADDITIVE v2 minor-revision key, omit-when-empty: absent on
   * every pre-motion document (and structurally impossible on legacy v1
   * wire), so the committed-baseline path never sees it. Components opt in
   * with an ordinary `AnimationName` property naming a key of this map;
   * a dangling reference is a defined no-op (§1.3), never an error.
   */
  keyframes?: IRKeyframes;
}

/**
 * One keyframe stop (spec 07 §1.2). Stops arrive from the converter
 * SORTED ascending by offset (stable for equal offsets — the authoring
 * order decides the css-animations-1 §4.2 last-wins cascade); readers may
 * rely on sortedness and must not reorder.
 */
export interface IRKeyframeStop {
  /** Resolved offset fraction in [0, 1] (`from`/`to`/percent pre-resolved). */
  offset: number;
  /** Typed stop declarations — the SAME {type, data} envelopes components use. */
  properties: IRProperty[];
}

/** Named keyframe sets: animation-name ident → sorted stop list. */
export type IRKeyframes = Record<string, IRKeyframeStop[]>;

/**
 * Child-side composition reference (schema/spec/03-children.md).
 *
 * Present on a component that previews inside a container; omitted for
 * roots. Structural data that MUST round-trip — engines never read it
 * (composition-agnostic contract); only composers (the harness) do.
 */
export interface IRSlot {
  /** `id` of the container this component composes into. */
  parent: string;
  /**
   * Slot name within the parent; reserved for future multi-slot
   * containers. The wire omits it at the default — `decodeIRDocument`
   * reconstructs `"content"` so consumers never see `undefined`.
   */
  name?: string;
}

/**
 * One entry of the wave-22 `meta.decorations` list — a single decorating
 * box's contribution to a COLLAPSED inline run (schema/spec/
 * 04-metadata-fields.md; producer: the `_decorations` banner in
 * tools/titan/extract-fixture.mjs).
 */
export interface IRDecoration {
  /**
   * Exactly ONE css-text-decor-3 §2.1 line keyword. An element declaring
   * two contributes two entries sharing one colour, so consumers never
   * re-tokenise. `none`/`blink` never appear (they paint nothing).
   */
  line: string;
  /**
   * The decorating box's `text-decoration-color` AS AUTHORED ('blue',
   * '#00f', 'rgb(0,0,255)') — NOT the normalized sRGB leaf every CSS
   * *property* value carries. On web that is a feature: the token goes
   * straight back into a `text-decoration-color` declaration and the
   * browser parses it, so the round trip is lossless by construction.
   * Absent/null = `currentColor` (§2.2 initial).
   */
  color?: string | null;
}

/**
 * One entry of the wave-32 `meta.runs` list — either an anonymous inline
 * text run or the position of one child box in the inline flow
 * (schema/spec/03-children.md §4.1; producer: the `_runs` banner in
 * tools/titan/extract-fixture.mjs).
 *
 * EXACTLY ONE key is set. Modelled as a union rather than
 * `{text?, child?}` so a malformed both-keys entry is a type error at
 * every consumer instead of a silent precedence question.
 */
export type IRRun =
  /** A bare text node at this position in the content walk. */
  | { text: string; child?: undefined }
  /**
   * The child at this position, named by its AUTHORING KEY — which is the
   * child's `name` on the v2 wire and (extractor-direct pipeline) also its
   * `id`. NOT the converter-minted id: the converter re-ids at the flatten
   * boundary, so an id written by the producer would name nothing after
   * the hop (schema/spec/04-metadata-fields.md).
   */
  | { child: string; text?: undefined };

/**
 * Droppable renderer hints, grouped (v2 home of v1's `_tag` / `_role`).
 * Omitted entirely when empty; strict (`{sourceTag?, role?, attrs?,
 * decorations?, markerText?, runs?}` only) when present — see
 * schema/spec/01-envelope.md.
 */
export interface IRMeta {
  /**
   * Originating HTML element tag (lowercase: 'ol', 'li', 'p', …), the v2
   * rename of v1 `_tag`. Lets the web renderer choose a matching native
   * element so browser-default semantics (list markers, paragraph
   * spacing, table layout) apply. Absent → render as <div>.
   */
  sourceTag?: string | null;
  /** Freeform authoring role hint (v2 rename of v1 `_role`). Droppable. */
  role?: string | null;
  /**
   * wave-20 widget-identity attributes (v2 home of the extractor's
   * `_attrs`). Emitted only for form/widget sourceTags (a/button/input/
   * textarea/select/option/meter/progress), only for present-in-source
   * attributes among {type, value, checked, multiple, size, alt, min,
   * max, selected, disabled}; booleans are presence-`true`, min/max —
   * and value on meter/progress — are numbers where numeric, everything
   * else verbatim source strings (schema/ir-v2.schema.json meta.attrs).
   * Application policy lives in renderer/WidgetAttrs.ts.
   */
  attrs?: Record<string, string | number | boolean> | null;
  /**
   * wave-22 per-line decorations for a COLLAPSED inline run (v2 home of
   * the extractor's `_decorations`). ORDER is OUTERMOST-FIRST — the
   * css-text-decor-3 §2.1 propagation order — and the list is
   * AUTHORITATIVE when present: it is the complete line set for the run,
   * so a renderer that honours it MUST ignore the component's own
   * `text-decoration-line` (the surviving flat longhands are a root-wins
   * merged bag kept for readers that drop `meta`). A present-but-EMPTY
   * list still says "no lines" and must paint nothing.
   * Application policy lives in renderer/DecorationSpans.ts.
   */
  decorations?: IRDecoration[] | null;
  /**
   * wave-27 resolved list-marker string for one `<li>` (v2 home of the
   * extractor's `_markerText`) — representation + suffix per
   * css-counter-styles-3 §6, already resolved against `<ol start>` and
   * the item's ordinal. Modelled so the type mirrors the wire, but the
   * web renderer DELIBERATELY DOES NOT CONSUME IT: it emits a real
   * `<ol>`/`<li>` and lets the browser synthesise `::marker` from the
   * component's `list-style-type` plus the forwarded `start` attribute,
   * which is a better oracle than any string we could bake. The two
   * native runtimes — which have no `::marker` — are the readers.
   */
  markerText?: string | null;
  /**
   * wave-32 ORDERED inline-content list (v2 home of the extractor's
   * `_runs`) — the component's own text and its kept children INTERLEAVED
   * in document order, the one shape the single `text` string cannot
   * express (`the quick <u>brown</u> fox` with a surviving `<u>`).
   *
   * AUTHORITATIVE when present: NodeRenderer paints the entries in order,
   * does NOT also paint `text`, and does NOT paint a referenced child a
   * second time from the sibling walk. Children this list does not
   * reference still render, after the runs, in sibling order (spec 03
   * §4.1 rule 4) — so a partial list can never make a box disappear.
   *
   * Each `{text}` entry is an INLINE anonymous run: a bare text node in
   * the children walk, never a wrapper element. A block box inside an
   * inline box splits it (CSS 2.1 §9.2.1.1), which would re-break the
   * very line box this key exists to preserve — the measured wave-31
   * "span lesson".
   */
  runs?: IRRun[] | null;
}

/**
 * A single UI component with its styles — v2 flat shape.
 *
 * NOTE: there is NO `children` field. Its presence in a v2 document is a
 * hard decode error (spec 03 §1); composition is rebuilt by the harness
 * composer from `slot` refs (Mode A) or supplied externally (Mode B).
 */
export interface IRComponent {
  /** Unique identifier across the WHOLE document (e.g. "button-001"). */
  id: string;
  /** Component type/class name (e.g. "Button", "Card"). */
  name: string;
  /** List of CSS properties as IR; ITEM-scoped placement claims included. */
  properties: IRProperty[];
  /**
   * CSS custom-property definitions declared on this component
   * (`--name` → RAW declaration value, verbatim). Additive IR v2 minor
   * revision (spec 01 component table). Names are case-SENSITIVE
   * (css-variables-1 §2) and values are untyped token streams until a
   * var() reference substitutes them; the empty string is a legal value.
   * var() resolution walks element → slot-parent chain → fallback
   * (spec 02 custom-properties section) and happens at style-build
   * time, never at decode time. Omitted when the component defines none.
   */
  variables?: Record<string, string>;
  /** State-based styles (hover, focus, …). v2 omits the key when empty. */
  selectors?: IRSelector[];
  /** Responsive breakpoint styles. v2 omits the key when empty. */
  media?: IRMedia[];
  /** Child-side composition ref; absent for roots (spec 03). */
  slot?: IRSlot | null;
  /**
   * Element text content (v2 rename of v1 `_text`). Empty string is a
   * legal emitted value ("extracted, was empty"); absent means "no text
   * — placeholder behaviour applies" so the 327-pair baseline is stable.
   */
  text?: string | null;
  /**
   * Pseudo-element payload (v2 rename of v1 `_pseudo`). Forwarded
   * verbatim from the authoring extractor (spec 01: extractor-owned,
   * never flattened into the component list) — hence the tolerant
   * {@link IRPseudoNode} inner shape.
   */
  pseudos?: PseudoElements | null;
  /** Grouped droppable hints; v2 home of v1 `_tag`/`_role`. */
  meta?: IRMeta | null;
}

/**
 * Per-element pseudo-element payload — see {@link IRComponent.pseudos}.
 *
 * Each slot is independently optional: a `<li>` might only carry a
 * `marker`, a `<span>` only a `before`. Renderers MUST treat missing/null
 * slots as "no synthetic node for that pseudo".
 */
export interface PseudoElements {
  /** Synthetic ::before node — rendered as a leading inline span. */
  before?: IRPseudoNode | null;
  /** Synthetic ::after node — rendered as a trailing inline span. */
  after?: IRPseudoNode | null;
  /** Synthetic ::marker node — rendered as a list-item-marker span. */
  marker?: IRPseudoNode | null;
}

/**
 * One synthetic pseudo-element node. The wire forwards the extractor's
 * authoring shape VERBATIM (spec 01: `pseudos` is extractor-owned), so
 * every field is optional and BOTH text spellings are tolerated: the
 * legacy `_text` (what the WPT extractor emits today) and a future
 * `text`. Style declarations from the original CSS rule (`content`,
 * color, font-*) arrive in `properties` and reach buildStyles intact.
 */
export interface IRPseudoNode {
  /** Optional id, used only for React keys / debug attributes. */
  id?: string;
  /** Optional debug name. */
  name?: string;
  /**
   * Style declarations from the originating pseudo rule. The payload is
   * extractor-owned and forwarded verbatim (spec 01), so BOTH shapes are
   * legal on the wire: a typed IR list, or the raw declarations map the
   * WPT extractor emits today ({ display: 'block', … }). NodeRenderer
   * bridges each shape to inline styles accordingly.
   */
  properties?: IRProperty[] | Record<string, string> | null;
  /** Literal `content:` string in the extractor's legacy spelling. */
  _text?: string | null;
  /** Literal `content:` string in the v2 spelling (tolerated). */
  text?: string | null;
}

/**
 * A CSS property in IR format — identical envelope in v1 and v2.
 *
 * Uses generic `unknown` data to handle all 550 property types flexibly;
 * specific handling is done in the per-property Extractors.
 */
export interface IRProperty {
  type: string;
  data: unknown;
}

/**
 * Pseudo-class selector styles (e.g., :hover, :focus). `condition` is
 * stored WITHOUT the leading colon (spec 01).
 */
export interface IRSelector {
  condition: string;
  properties: IRProperty[];
}

/**
 * Media query styles (e.g., min-width: 768px). `query` is the raw media
 * query string, parentheses included (spec 01).
 */
export interface IRMedia {
  query: string;
  properties: IRProperty[];
}

File: runtimes/compose/src/main/java/com/styleconverter/runtime/core/ir/IRModels.kt
package com.styleconverter.runtime.core.ir

import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject

/**
 * Root document containing all components.
 *
 * @property keyframes Document-level named @keyframes sets (additive IR v2
 *   minor revision — schema/spec/07-animations.md §1.2). Keys are the
 *   lowercased custom-ident names `AnimationName` references; values are
 *   the offset-sorted stop lists. null when the wire omitted the key
 *   (omit-when-empty rule) and for every v1 document — the legacy path
 *   never carried keyframes, so the default keeps v1 decode byte-stable.
 * @property fontFaces Document-level `@font-face` declarations (additive IR
 *   v2 minor revision — schema/spec/01-envelope.md §5). null when the wire
 *   omitted the key (omit-when-empty rule) and for every v1 document.
 *
 *   DECODED BUT NOT YET CONSUMED, and that is the honest state as of wave
 *   34: Compose selects one face per weight/style through
 *   `FontFamily(Font(...))` and has no runtime hook to register an arbitrary
 *   file into the resolver, so every component referencing a declared family
 *   still renders in the bundled Inter. The field exists so the wire is not
 *   a web-only dialect and so the future registration hop reads the file
 *   path off the document instead of re-deriving it — `Typeface.Builder`
 *   takes exactly this string joined to the harness's asset root.
 */
@Serializable
data class IRDocument(
    val components: List<IRComponent>,
    val keyframes: Map<String, List<IRKeyframeStop>>? = null,
    val fontFaces: List<IRFontFace>? = null
)

/**
 * One document-level `@font-face` declaration (schema/spec/01-envelope.md §5).
 *
 * @property family The `font-family` descriptor (css-fonts-4 §4.2), already
 *   UNQUOTED by the writer — matched against the family names in an
 *   element's `font-family` value.
 * @property src Path to the font FILE, relative to the producing pipeline's
 *   corpus root. A path, never a payload (spec 01 §5 carries the size
 *   argument); the writer guarantees the file existed at emit time.
 * @property weight The `font-weight` descriptor AS AUTHORED (§4.4) — may be
 *   a RANGE ("400 700"), which is why this is a String and not the IR's
 *   numeric 100–900 form. null = the §4.4 initial `normal`.
 * @property style The `font-style` descriptor AS AUTHORED (§4.5) — may carry
 *   an oblique angle ("oblique 20deg"). null = the §4.5 initial `normal`.
 */
@Serializable
data class IRFontFace(
    val family: String,
    val src: String,
    val weight: String? = null,
    val style: String? = null
)

/**
 * One resolved keyframe stop (spec 07 §1.2).
 *
 * @property offset The RESOLVED fraction in [0, 1] — the converter already
 *   mapped `from`/`to`/percent selectors, so readers never re-parse them.
 *   Stops arrive sorted ascending (stable for equal offsets); readers MAY
 *   rely on sortedness and MUST NOT reorder.
 * @property properties Standard typed `{type, data}` property envelopes —
 *   byte-identical semantics to component properties, including the
 *   unknown-type tolerance rule (skip + log at use, never at decode).
 */
@Serializable
data class IRKeyframeStop(
    val offset: Double,
    val properties: List<IRProperty> = emptyList()
)

/**
 * The IR v2 child→parent composition reference (schema/spec/03-children.md).
 *
 * Carried by a CHILD component in a flat v2 document; roots omit it.
 * `slot` is STRUCTURAL data that MUST round-trip (spec 05 tolerance rule 4)
 * — unlike the droppable `meta` hints. Only composers (the harness's
 * SlotComposer, later an SDUI shell) may read it; the style engine itself
 * is composition-agnostic by contract and never consults slot.
 *
 * @property parent The `id` of the container this component previews inside.
 * @property name Slot name within the parent; the wire omits it when it
 *   equals the documented default "content" (reserved for future
 *   multi-slot containers — scaffold header/body/footer).
 */
@Serializable
data class IRSlot(
    val parent: String,
    val name: String = "content"
)

/**
 * Wave-20 wire contract (lane W2) — the form/widget attribute capsule the
 * extractor emits as `meta.attrs` for the widget tags (a, button, input,
 * textarea, select, option, meter, progress). ONLY present-in-source
 * attributes among the pinned ten appear; types are pinned by the wire
 * contract: strings except checked/multiple/selected/disabled (booleans)
 * and min/max/value-on-meter-progress (numbers where numeric — hence the
 * separate [valueNumber] channel so a numeric wire `value` survives
 * without stringly re-parsing). Droppable meta hint like `sourceTag`:
 * ignoring it loses fidelity, never correctness.
 */
@Serializable
data class IRAttrs(
    val type: String? = null,      // input type ("checkbox", "range", …)
    val value: String? = null,     // string-form value attribute
    val valueNumber: Double? = null, // numeric wire value (meter/progress)
    val checked: Boolean? = null,  // checkbox/radio checked presence
    val multiple: Boolean? = null, // select multiple presence (listbox)
    val selected: Boolean? = null, // option selected presence
    val disabled: Boolean? = null, // disabled presence (not yet painted)
    val size: String? = null,      // select/input size attribute
    val alt: String? = null,       // image-input alt text
    val min: Double? = null,       // meter/progress/range min
    val max: Double? = null,       // meter/progress/range max
    // wave-27 lane CBAKE: `<ol start>` — the ordered list's counter origin
    // (HTML §4.4.5), a VERBATIM string like every member of the disjoint
    // ol/li attr lane. Appended LAST so the wave-20 widget field order (and
    // every positional construction of it) is untouched. Decoded for wire
    // completeness; Compose paints the baked [IRComponent.markerText]
    // instead of counting from it. Twin of Swift IRAttrs.start.
    val start: String? = null,
    // wave-36 lane M1: the REPLACED-ELEMENT SOURCE — a third disjoint attr
    // lane (img/embed/object/video). ONE canonical key whatever the markup
    // spelled (`src` on img/embed, `data` on object per HTML §4.8.7,
    // `poster` on video per §4.8.9), carrying a PRODUCER-RELATIVE PATH or a
    // `data:` URI — never a payload — on the same consumer-resolves
    // contract as the document-level `fontFaces[].src`
    // (schema/spec/04-metadata-fields.md).
    //
    // DECODED, NOT YET PAINTED. Compose has no asset origin for a
    // producer-relative corpus path (a device cannot read the host's disk —
    // the same asymmetry the @font-face channel documents), so this field
    // exists so the strict v2 envelope ACCEPTS the key rather than throwing
    // on a wire the web consumer needs. Painting it is the named follow-up:
    // a bundling hop like the two feeders' --wpt-dir copy, then an
    // AsyncImage/Painter in ComponentRenderer. Appended LAST for the same
    // positional-construction reason `start` was. Twin of Swift IRAttrs.src.
    val src: String? = null
)

/**
 * Wave-22 wire contract (lane DECOR) — ONE entry of the `meta.decorations`
 * list a COLLAPSED inline run carries (schema/spec/04-metadata-fields.md;
 * producer: the `_decorations` banner in tools/titan/extract-fixture.mjs).
 *
 * The list is ORDERED OUTERMOST-FIRST — the css-text-decor-3 §2.1
 * propagation order, so a descendant's line paints over its ancestors'
 * where both land on the same row — and AUTHORITATIVE when present: it is
 * the complete line set for the run, so the painter ignores the
 * component's own `text-decoration-line` flags there.
 *
 * Both members stay RAW WIRE STRINGS on purpose. [line] is normally one
 * of the three §2.1 keywords, but the DECODER DOES NOT VALIDATE THE VALUE
 * — an unknown keyword is the spec-05 tolerance-rule-1 case and survives
 * decode raw, so the PAINT-time filter in DecorationWire can drop + log
 * it without collapsing a present list back to "absent" (IRDocumentDecoder
 * .decodeDecorations says the same; the iOS twin's doc comment matches).
 * [color] is the CSS colour token AS AUTHORED ("blue", "#00f",
 * "rgb(0,0,255)") — the converter does not normalize it to the IR sRGB
 * leaf (see 04-metadata-fields.md for why), so resolution happens in
 * DecorationWire.toDecorationLines via the runtime's own token parser.
 * A null [color] means the PAINTER substitutes — see that function for
 * what the substitute actually is (NOT plain currentColor: both painters
 * fall back to the run's merged `text-decoration-color`, else the text
 * colour).
 */
@Serializable
data class IRDecoration(
    val line: String,
    val color: String? = null
)

/**
 * A single UI component with its styles.
 *
 * @property id Unique identifier for SDUI (e.g., "button-001")
 * @property name Component type/class name (e.g., "Button", "Card")
 * @property properties List of CSS properties as IR
 * @property selectors State-based styles (hover, focus, etc.)
 * @property media Responsive breakpoint styles
 * @property children Nested child components for containers. v1 wire only:
 *   IR v2 documents are FLAT (children on the wire is a hard decode error —
 *   see IRDocumentDecoder); in-memory the field is (re)populated by
 *   SlotComposer from the v2 `slot` refs so the renderer sees one tree
 *   shape regardless of wire version.
 * @property _text Optional inner-text content carried from the WPT
 *   extractor. Renderer draws this verbatim instead of the placeholder
 *   name when present; also rendered as a leading inline node alongside
 *   children for mixed-content layouts (Bug 1, swarm-002 css-text-decor
 *   text-decoration-decorating-box-thickness-001). v2 wire name: `text`
 *   (IRDocumentDecoder maps it onto this field; the in-memory name keeps
 *   the v1-era spelling so the renderer and its pinning tests are
 *   untouched by the wire rename).
 * @property _tag Optional originating HTML element tag (lowercase: 'ol',
 *   'li', 'p', etc.). Lets the renderer wire up tag-default behaviour —
 *   list-marker prefix on <li> children of <ol>/<ul>, paragraph spacing
 *   on <p>, table layout on <tr>/<td>, etc. (Bug 2, swarm-002
 *   css-counter-styles css3-counter-styles-101). Both fields are
 *   nullable so legacy fixtures (visual-test.json) deserialize
 *   unchanged and the 327-pair baseline stays stable. v2 wire name:
 *   `meta.sourceTag`.
 * @property slot v2 composition reference (see [IRSlot]); null for roots
 *   and for every component of a v1 document.
 * @property pseudos v2 generated-content payload (wire rename of the
 *   authoring `_pseudo`): an opaque component-shaped {before?, after?,
 *   marker?} object. Pseudo nodes never flatten (design §4.2) — they have
 *   no independent lifecycle. Decoded and retained so the contract
 *   round-trips; the Compose renderer does not consume it yet (its
 *   ::before/::after support still rides the selectors channel — TODO
 *   route pseudos into ContentApplier when the web/iOS ordering contract
 *   is pinned cross-platform).
 * @property role v2 `meta.role` droppable hint (only value emitted today
 *   is "body-root"). v1 wire spelled it `_role` and this model dropped it
 *   (the documented spec-04 caveat); the v2 decoder closes that gap.
 * @property attrs Wave-20 widget-identity capsule (see [IRAttrs]) — v2
 *   wire name `meta.attrs`, riding the meta channel exactly like `_tag`
 *   (the documented precedent). Null for every v1 document and for
 *   components the extractor didn't tag (non-widget elements).
 * @property decorations Wave-22 per-line decoration list (see
 *   [IRDecoration]) — v2 wire name `meta.decorations`, riding the meta
 *   channel exactly like `attrs`. Null for every v1 document and for every
 *   run the extractor did not collapse; an EMPTY list is a distinct,
 *   meaningful state ("authoritative and it says: no lines") that the
 *   painter must honour by drawing nothing.
 * @property markerText Wave-27 RESOLVED list-marker string for one `<li>`
 *   — v2 wire name `meta.markerText`, riding the meta channel exactly like
 *   `attrs`. Produced by the extractor's counter-style bake, which owns the
 *   whole css-counter-styles-3 §6 resolution (predefined table, §4 range,
 *   §7.1.4 fallback, §3.1.5 suffix) plus the `<ol start>` / `<li value>`
 *   ordinal the IR has no property for. AUTHORITATIVE when present: the
 *   renderer must render it INSTEAD of calling
 *   [com.styleconverter.runtime.lists.StyleListApplier.getMarker], never
 *   as well. Null for every v1 document, every non-`<li>` component, and
 *   every marker family the bake leaves to this runtime's own table
 *   (the §6.1 bullets, `none`, and unmodelled counter styles).
 * @property lang Wave-37 COMPUTED content language of this component's
 *   source element — v2 wire name `meta.lang`, riding the meta channel
 *   exactly like `attrs`. ALREADY RESOLVED by the producer through HTML
 *   §3.2.6.2's own-lang → nearest-ancestor → `<html>`/`<body>` ladder,
 *   because the flat v2 component list has no parent edge this decoder
 *   could walk. VERBATIM as authored ("eN-Us" stays "eN-Us"): RFC 4647
 *   matching is case-insensitive and subtag-truncating, so a consumer
 *   lowercases at LOOKUP, never at decode. Null when the document declares
 *   no language, which is the same state as "use the default locale" —
 *   the state every pre-wave-37 document is in.
 * @property variables CSS custom-property definitions declared on this
 *   component ("--name" → RAW declaration value, verbatim). Additive IR
 *   v2 envelope key (schema/spec/01-envelope.md): names are
 *   case-SENSITIVE and values stay untyped token streams until var()
 *   substitution (css-variables-1 §2). null when the wire omitted the
 *   key. Resolution (element → slot-parent chain → fallback →
 *   guaranteed-invalid, schema/spec/02-values.md) is the style engine's
 *   job — the decoder only round-trips the map.
 */
/**
 * Wave-32 wire contract (lane R) — ONE entry of the `meta.runs` list an
 * INTERLEAVING component carries (schema/spec/03-children.md §4.1;
 * producer: the `_runs` banner in tools/titan/extract-fixture.mjs).
 *
 * `_text` is ONE string and the child list has ONE order, so the wire could
 * say `run + children` or `children + run` but never `run / child / run` —
 * the shape `the quick <u>brown</u> fox` needs when the `<u>` survives as a
 * child. `meta.runs` is that ordering, and ORDER IS THE WHOLE PAYLOAD.
 *
 * EXACTLY ONE member is non-null, enforced at decode
 * (IRDocumentDecoder.decodeRuns) rather than by the type, because the wire
 * shape is `{text}` | `{child}` and a Kotlin sealed hierarchy would not
 * round-trip it any more safely than this pair does.
 *
 * [child] is the referenced child's AUTHORING KEY — its `name` on the
 * converter-emitted wire, and also its `id` in the extractor-direct
 * pipeline. NOT the converter-minted id: the converter re-ids at the
 * flatten boundary, so an id written by the producer would name nothing
 * after the hop (schema/spec/04-metadata-fields.md).
 */
@Serializable
data class IRRun(
    val text: String? = null,
    val child: String? = null
)

@Serializable
data class IRComponent(
    val id: String,
    val name: String,
    val properties: List<IRProperty> = emptyList(),
    val selectors: List<IRSelector> = emptyList(),
    val media: List<IRMedia> = emptyList(),
    val children: List<IRComponent>? = null,
    val _text: String? = null,
    val _tag: String? = null,
    val attrs: IRAttrs? = null,
    val decorations: List<IRDecoration>? = null,
    val markerText: String? = null,
    // Wave-37 (lane W4): the element's COMPUTED content language
    // (`meta.lang`) — see the @property doc above. Droppable: a renderer
    // that ignores it paints in the default locale, which costs LOCALE
    // fidelity (CLDR quote pairs, generic-family font fallback) and never
    // correctness.
    val lang: String? = null,
    // Wave-32 (lane R): the ORDERED inline-content list (`meta.runs`) — the
    // component's own text and its kept children INTERLEAVED in document
    // order. AUTHORITATIVE when present: RenderContent paints the entries in
    // order and must NOT also paint [_text], nor paint a referenced child a
    // second time. Null for every component whose text does not glue across
    // a child, which is all but ~1,138 of the committed corpus — see
    // [IRRun] and schema/spec/03-children.md §4.1.
    val runs: List<IRRun>? = null,
    val slot: IRSlot? = null,
    val pseudos: JsonObject? = null,
    val role: String? = null,
    val variables: Map<String, String>? = null
)

/**
 * A CSS property in IR format.
 *
 * Uses generic JsonElement for data to handle all 446+ property types flexibly.
 * Specific property handling is done in StyleApplier.
 */
@Serializable
data class IRProperty(
    val type: String,
    val data: JsonElement
)

/**
 * Pseudo-class selector styles (e.g., :hover, :focus).
 */
@Serializable
data class IRSelector(
    val condition: String,
    val properties: List<IRProperty>
)

/**
 * Media query styles (e.g., min-width: 768px).
 */
@Serializable
data class IRMedia(
    val query: String,
    val properties: List<IRProperty>
)
